### ISSUE TITLE
camera->supportedViewfinderSettings().size() == 0 on Android

### ISSUE THREAD

**Issue by headupinclouds** [headupinclouds/gatherer#109]
Need to investigate so we can configure max resolution on Android devices.

**Comment by headupinclouds** [comment]
https://bugreports.qt.io/browse/QTBUG-46470

**Comment by headupinclouds** [comment]
Perhaps something like section A, B, or C below?  This doesn't work on iOS but seems to have some effect on Android, but it doesn't provide the exact requested resolution.  

```
    {
        std::pair<QSize, int> best;
        QMediaRecorder *videoCapture = new QMediaRecorder(camera);
        QCameraImageCapture *imageCapture = new QCameraImageCapture(camera);
        QList<QVideoFrame::PixelFormat> formats = imageCapture->supportedBufferFormats();

        QList<QSize> resolutions = imageCapture->supportedResolutions();
        //QList<QSize> resolutions = videoCapture->supportedResolutions(); // recording only?
        if(resolutions.size())
        {
            // This seems to work on Android, but not for iOS
            for(auto &i : resolutions)
            {
                int area = i.width() * i.height();
                if(area > best.second)
                {
                    best = {i, area};
                }
                logger->info() << "video: " << i.width() << " " << i.height();
            }

            logger->info() << "best: " << best.first.width() << " " << best.first.height();

            // (A)
            QVideoEncoderSettings videoEncoderSettings;
            videoEncoderSettings.setResolution(best.first);
            videoCapture->setVideoSettings(videoEncoderSettings);

            // (B)
            QImageEncoderSettings imageSettings;
            imageSettings.setResolution(best.first);
            imageCapture->setEncodingSettings(imageSettings);

            // (C)
            QCameraViewfinderSettings settings;
            settings.setResolution(best.first);
#if defined(Q_OS_IOS)
            settings.setPixelFormat(QVideoFrame::Format_NV12);
#else
            settings.setPixelFormat(QVideoFrame::Format_NV21);
#endif
            camera->setViewfinderSettings(settings);
            auto result = camera->viewfinderSettings();
        }
    }
```

This gives the following output on an Android device, but I'm not able to configure 2560x1440 resolution.

```
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 320 240
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 640 480
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 720 480
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 960 720
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 1280 720
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 1280 960
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 1440 1080
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 1920 1080
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 1440 1440
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 2560 1080
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 1920 1440
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: video: 2560 1440
I/spdlog  (31506): [15:47:19.965 | thread:31529 | qmlvideofilter | info]: best: 2560 1440
```

**Comment by ruslo** [comment]
> but I'm not able to configure 2560x1440 resolution

Do you have some runtime error or you just see that resolution is lower?

Just for your information I have:

```
I/spdlog  (14930): [23:37:18.842 | thread:14944 | qmlvideofilter | info]: Start
I/spdlog  (14930): [23:37:19.586 | thread:14944 | qmlvideofilter | info]: video: 176 144
I/spdlog  (14930): [23:37:19.587 | thread:14944 | qmlvideofilter | info]: video: 320 240
I/spdlog  (14930): [23:37:19.587 | thread:14944 | qmlvideofilter | info]: video: 352 288
I/spdlog  (14930): [23:37:19.587 | thread:14944 | qmlvideofilter | info]: video: 640 480
I/spdlog  (14930): [23:37:19.587 | thread:14944 | qmlvideofilter | info]: best: 640 480
```

Application works fine but I'm not sure what resolution set for camera in fact (I see only 640x480 value in Qt's widget).

**Comment by headupinclouds** [comment]
> Do you have some runtime error or you just see that resolution is lower?

The application runs okay, but it is unclear how to change the camera resolution, or which resolutions are valid.  (It isn't clear to me that there is necessarily a problem, but I have yet to find an explanation for which video resolutions are supported or how to configure those video resolutions properly.)

This was just the result of a few quick experiments based on some of the QT code snippets I had found from related posts.  I haven't tracked down the corresponding code.

It does seem to be a known limitation that Android doesn't implement `supportedViewfinderSettings`.  That seems to leave either `QMediaRecorder::setVideoSettings(QVideoEncoderSettings...)` or `QCameraImageCapture:: setEncodingSettings(QImageEncoderSettings...)`.  It seems resolutions for `QCameraImageCapture` are for the still photo captures and `QMediaRecorder` resolutions are for recorded videos.  It isn't clear to me what the relationship is between live preview/video resolutions and these two classes/methods.

Since you are already seeing maximum resolution on your device I think you would need to check if you can configure the video for any of the lower resolutions based on reported video or image resolutions (see below).  Note that the logged resolutions currently correspond to the "image" settings.  You can switch between image and video resolutions here:

```
        QList<QSize> resolutions = imageCapture->supportedResolutions();
        //QList<QSize> resolutions = videoCapture->supportedResolutions(); // recording only?
```

And then use one of `videoCapture->setVideoSettings(videoEncoderSettings);`  `imageCapture->setEncodingSettings(imageSettings);` or `camera->setViewfinderSettings(settings);`.  And there is an option to display the configured video resolution using `auto result = camera->viewfinderSettings();`.  I'm not sure at what point this is valid/initialized, or if it requires starting the camera up for this to take effect.

**Comment by ruslo** [comment]
Just for the record I've checked that `QCamera::supportedViewfinderSettings` returns from here:
- https://github.com/hunter-packages/Qt/blob/84a5846a20da62e2e2c73c1218927f8746c42de6/qtmultimedia/src/multimedia/camera/qcamera.cpp#L652

it means that for Android `QCameraViewfinderSettingsControl2` can't be found:
- https://github.com/hunter-packages/Qt/blob/84a5846a20da62e2e2c73c1218927f8746c42de6/qtmultimedia/src/multimedia/camera/qcamera.cpp#L185

>  https://bugreports.qt.io/browse/QTBUG-46470

Don't understand why this bug is closed. Developer mentioned that viewfinder settings is not implemented on Android but resolution is `Out of scope`. I've openned new feature request https://bugreports.qt.io/browse/QTBUG-50422 for tracking/discussion purposes.

**Comment by ruslo** [comment]
> And there is an option to display the configured video resolution using auto result = camera->viewfinderSettings();

Both `d->viewfinderSettingsControl2` and `d->viewfinderSettingsControl` is `nullptr` so nothing will be returned:
- https://github.com/hunter-packages/Qt/blob/84a5846a20da62e2e2c73c1218927f8746c42de6/qtmultimedia/src/multimedia/camera/qcamera.cpp#L557

**Comment by ruslo** [comment]
> imageCapture->setEncodingSettings(imageSettings)

Works for me, I've choosed second best resolution:

```
I/spdlog  ( 7206): [01:39:03.896 | thread:7221 | qmlvideofilter | info]: video: 176 144
I/spdlog  ( 7206): [01:39:03.896 | thread:7221 | qmlvideofilter | info]: video: 320 240
I/spdlog  ( 7206): [01:39:03.896 | thread:7221 | qmlvideofilter | info]: video: 352 288
I/spdlog  ( 7206): [01:39:03.896 | thread:7221 | qmlvideofilter | info]: video: 640 480
I/spdlog  ( 7206): [01:39:03.896 | thread:7221 | qmlvideofilter | info]: best: 352 288
```

And size of the VideoFrame receiving by FilterRunnable changed.

**Comment by ruslo** [comment]
> videoCapture->setVideoSettings(videoEncoderSettings)

Doesn't work for me

**Comment by headupinclouds** [comment]
I'm told by `imageCapture->supportedResolutions()` that these resolutions are available:

```
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 320 240
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 640 480
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 720 480
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 960 720
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 1280 720
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 1280 960
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 1440 1080
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 1920 1080
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 1440 1440
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 2560 1080
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 1920 1440
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: video: 2560 1440
I/spdlog  (26788): [12:07:48.724 | thread:26815 | qmlvideofilter | info]: best: 2560 1440
```

I try to request the highest resolution (`2560 1440`) using the `imageCapture->setEncodingSettings(imageSettings)` option for an android device, but I actually get `1920x1080` video output.  I'm guessing these are all valid photo resolutions, but that there is some subset of valid video/preview resolutions.  My guess was that might be the same as the `videoCapture->supportedResolutions()` list:

```
I/spdlog  (28771): [12:14:48.424 | thread:28787 | qmlvideofilter | info]: video: 176 144
I/spdlog  (28771): [12:14:48.424 | thread:28787 | qmlvideofilter | info]: video: 320 240
I/spdlog  (28771): [12:14:48.424 | thread:28787 | qmlvideofilter | info]: video: 720 480
I/spdlog  (28771): [12:14:48.424 | thread:28787 | qmlvideofilter | info]: video: 1280 720
I/spdlog  (28771): [12:14:48.424 | thread:28787 | qmlvideofilter | info]: video: 1920 1080
I/spdlog  (28771): [12:14:48.424 | thread:28787 | qmlvideofilter | info]: best: 1920 1080
```

**Comment by ruslo** [comment]
I can reproduce issue by changing Front to Back camera - best resolution is 2592x1944 but frame returns 1280x720.

**Comment by ruslo** [comment]
`imageCapture->supportedResolutions` is closer to the truth but not all of the listed formats are supported in fact. Returned list:
- 176x144 (fail)
- 320x240 (fail)
- 352x288 (ok)
- 720x480 (ok)
- 1280x720 (ok)

**Comment by headupinclouds** [comment]
From a [SO post](http://stackoverflow.com/a/5934967/5724090): "Did you try to use the `getSupportedVideoSizes()` method from Camera.Parameters class?"

We could probably try this in : [OGVideoProcDroid](https://github.com/headupinclouds/ogles_gpgpu/tree/hunter/examples/android/OGVideoProcDroid)

I'm not sure what the expected behavior is for all of this  in QT.   There is an option for checking the resolution `auto result = camera->viewfinderSettings()` that could be used after calls to `imageCapture->setEncodingSettings(imageSettings)` which might tell us which ones work.

**Comment by ruslo** [comment]
>  There is an option for checking the resolution auto result = camera->viewfinderSettings()

Will show us nothing, see previous comment: https://github.com/headupinclouds/gatherer/issues/109#issuecomment-171003387

**Comment by ruslo** [comment]
> We could probably try this in : OGVideoProcDroid

So the plan is to call native Java code from C++? I think it's easier to figure out how to implement it in Qt. See for example: https://github.com/hunter-packages/Qt/blob/hunter-5.5.1/qtmultimedia/src/plugins/android/src/wrappers/jni/androidcamera.cpp#L802

**Comment by headupinclouds** [comment]
> So the plan is to call native Java code from C++

Sorry, that wasn't clear.  I just mean we could verify what the supported preview/video resolutions are in an existing Android app (and in Android docs) using the Android/Java/Camera API.  In my case it might be that `2560x1440` just isn't a supported video resolution, even from the Java layer.  The missing piece in QT seems to be a clear definition of how to obtain valid/supported video resolutions.

**Comment by ruslo** [comment]
>  I just mean we could verify what the supported preview/video resolutions are in an existing Android app

Still don't get it. We can use "hacky" `imageCapture->setEncodingSettings` and hope that it will set correct parameters. To check the parameters we need to call `Java/Camera API` from C++, right? Then it make sense to investigate how this is done in Qt, probably it will be easier to fix this bug then introduce Java-C++ layer from scratch.

**Comment by ruslo** [comment]
Summary from Tox chat: "Patch Java ogles example to test native Android Camera API. Resolutions. Back/front camera."

**Comment by ruslo** [comment]
> Patch Java ogles example to test native Android Camera API. Resolutions. Back/front camera

Variables for changing resolutions marked with `TESTME`: https://github.com/headupinclouds/ogles_gpgpu/pull/13/files

Front/Back cameras can be choosed here: https://github.com/headupinclouds/ogles_gpgpu/blob/19761db8f91105e5b601c99f61e2338c91ec804c/examples/android/OGVideoProcDroid/src/ogles_gpgpu/examples/ogvideoprocdroid/CameraProcThread.java#L363

For now changes can be done in Java code. In future may be UI widgets can be used.

**Comment by ruslo** [comment]
I've tested ogles example, works as expected. Checks (both front and back cameras):
- subjective: 1280x960 looks much better than 176x144
- semi-subjective: I've made screenshots with all available resolutions and view them on my laptop - it's a sequence of similar images from best to worst
- objective: `getPreviewSize` returns the exact value that is set by `setPreviewSize`

**Comment by ruslo** [comment]
List of resolutions on my Android device returned by native Camera API.

# Back facing camera

1280x960 OK (returned by 'getPreferredPreviewSizeForVideo')
1280x720 OK
864x480 OK
800x480 OK
768x432 OK
720x480 OK
640x480 OK
576x432 OK
480x320 OK
384x288 OK
352x288 OK
320x240 OK
240x160 OK
176x144 OK

# Front facing camera

640x480 OK (returned by 'getPreferredPreviewSizeForVideo')
576x432 OK
480x320 OK
384x288 OK
352x288 OK
320x240 OK
240x160 OK
176x144 OK

**Comment by ruslo** [comment]
Front camera resolutions returned by `QCameraImageCapture::supportedResolutions`:
- 176x144
- 320x240
- 352x288
- 640x480

Front camera resolutions returned by `QMediaRecorder::supportedResolutions`:
- 176x144
- 320x240
- 352x288

So none of these lists matches native "real" list of available resolutions.

**Comment by ruslo** [comment]
From Tox chat: will see how Java native Android API can be called from C++ using Qt's wrappers. Effectively investigate fix of the `viewfinderSettings` for Android ([bug #50422](https://bugreports.qt.io/browse/QTBUG-50422)).

**Comment by ruslo** [comment]
http://lists.qt-project.org/pipermail/development/2016-February/024737.html

**Comment by ruslo** [comment]
Stuck with C++/Java array magic: http://forum.qt.io/topic/63844/calling-from-c-java-method-which-returns-list-int

**Comment by ruslo** [comment]
Patched version doesn't work as I expect when I've tried to change resolution. Here are some places I have question about:
- https://github.com/hunter-packages/Qt/blob/4c9f6c083dbb33732dc4a8a064936946873cf0b9/qtmultimedia/src/plugins/android/src/mediacapture/qandroidcamerasession.cpp#L289

We are setting preview size according to the aspect value but not the exact match. E.g. I'm trying to set 576x432 (aspect ratio = 4:3 ~ 1.33) but since 640x480 (aspect ratio = 4:3 ~ 1.33) is supported we pick 640x480.

I've added debug message and discover that `QAndroidCameraSession::startPreview` called:
- https://github.com/hunter-packages/Qt/blob/4c9f6c083dbb33732dc4a8a064936946873cf0b9/qtmultimedia/src/plugins/android/src/mediacapture/qandroidcamerasession.cpp#L389

Later this function do call `adjustViewfinderSize` again:
- https://github.com/hunter-packages/Qt/blob/4c9f6c083dbb33732dc4a8a064936946873cf0b9/qtmultimedia/src/plugins/android/src/mediacapture/qandroidcamerasession.cpp#L412

but with resolution taken from `m_imageSettings` which is not set to 576x432, effectively rewriting user's intention to set custom resolution.

**Comment by ruslo** [comment]
I've created branch `android.viewfinder` with experimental Qt version: https://github.com/headupinclouds/gatherer/tree/android.viewfinder

**Comment by ruslo** [comment]
Fixed by #138
